### PR TITLE
Design: argo workflow on SQLFlow

### LINKED ISSUE #1423: Need design doc for pkg/argo
I guess we are going to have two APIs to (1) submit workflows to Argo for executions, and (2) to retrieve the log by each execution ID.

But by reading the current source code, I am not sure if it is what we need.  For example, I don't understand why would we need two functions with very similar names `isCompletedPhaseWF` and `isCompletedPhasePB`. Indeed, I don't get what does it mean `WF` or `PB`.  In particular, there is a function `translatePhase` which seems translating the input type of `isCompletedPhaseWF` into that of `isCompletedPhasePB`.

Another confusion is that I'd expect `func Fetch(token pb.FetchToken) (*pb.FetchResponse, error)` to be an asynchronous operation using Goroutine and Go channels.  It doesn't. And there is a comment there saying `// FIXME(tony): no need to wait for the whole workflow`.

I think we need to follow our convention -- to write the design before the code.  Let's list the API functions we need and how they are supposed to be used.

### PR DESCRIPTION
**Pull request by Yancey0623**
fixed #1423 
Written this design doc with @tonyyang-svail

### CODE REVIEW

File: doc/design/argo_workflow_on_sqlflow.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+```
+
+## Usage
+
+We expect the user to call `Submit` once then keep calling `Fetch` until completion. For example,
+
+```go
+func WorkflowDemo() {
+  wfJob := Submit(argoYAML)
+  token := newFetchToken(wfJob)
```
Comment (wangkuiyi) [inline doc/design/argo_workflow_on_sqlflow.md:55]: Is it a token or an ID?

File: doc/design/argo_workflow_on_sqlflow.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+When SQLFLow server receives a gRPC `Run` request that contains a SQL program, it:
+
+1. Translates the SQL program to an Argo YAML file.
+1. Submits the YAML file to Kubernetes and receives an Argo workflow ID. `<---(1)`
+1. Returns the workflow ID as a job token to the client.
+
+When SQLFLow server receives a gRPC `Fetch` request contains a job token, it:
+
+1. Looks up the associated Argo workflow and fetches most recent logs. `<---(2)`
```
Comment (wangkuiyi) [inline doc/design/argo_workflow_on_sqlflow.md:15]: How to define "most recent logs"?

Suppose there are two successive calls to  `Fetch`.  The first call returns logs generated during the period from the starting of the workflow to moment A. How could we implement `Fetch` to make sure that the second call to it returns logs generated after moment A?  Does the SQLFlow server need to keep track of moment A for each workflow ID?  If so, the SQLFlow server keeps workflow status, which is exactly what we DON'T want and is the reason we introduce Argo, an execution engine to run workflows.

**Review by wangkuiyi — CHANGES_REQUESTED** [review]

File: doc/design/argo_workflow_on_sqlflow.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+We design `Fetch` as `Fetch(token FetchToken) FetchResponse` where
+
+- `FetchToken` contains the workflow ID and the current fetching state.
+- `FetchResponse` contains the most recent logs and refreshed `FetchToken`.
+
+We implement the `Fetch` API as the following logic:
+
+``` go
+func Fetch(token FetchToken) *FetchResponse {
```
Comment (typhoonzero) [inline doc/design/argo_workflow_on_sqlflow.md:37]: How about link to https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/design/high_available_sqlflow_server.md or restruct these two design docs?

File: doc/design/argo_workflow_on_sqlflow.md (hunk 1 above)
Comment (Yancey0623) [reply doc/design/argo_workflow_on_sqlflow.md:55]: The `Token` contains `ID` and fetch status such as `logOffset` and `stepId`:

https://github.com/sql-machine-learning/sqlflow/blob/36066f128f99acc23e47f6fb93e9d5a6081fad5e/pkg/proto/sqlflow.proto#L52-L58

File: doc/design/argo_workflow_on_sqlflow.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  newToken := updateToken(token, newLogOffset, isFinished)
+
+  return newFetchReponse(logs, newToken)
+}
+```
+
+## Usage
+
+We expect the user to call `Submit` once then keep calling `Fetch` until completion. For example,
```
Comment (Yancey0623) [inline doc/design/argo_workflow_on_sqlflow.md:50]: Move `usage` content to Motivation.

File: doc/design/argo_workflow_on_sqlflow.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+We implement the `Fetch` API as the following logic:
+
+``` go
+func Fetch(token FetchToken) *FetchResponse {
+  pod, logOffset := getPodAndLogOffset(token)
+
+  logs, newLogOffset, isFinished := FetchPodLogs(pod, logOffset)
+
+  newToken := updateToken(token, newLogOffset, isFinished)
```
Comment (Yancey0623) [inline doc/design/argo_workflow_on_sqlflow.md:42]: Add some details about `updateToken`

**Review by wangkuiyi — APPROVED** [review]
I am approving the merge of this PR. But I think we will have some following PRs.

1. The structure of this document should be 
   1. Motivations -- we want to design a mechanism like how Google search engine returns multiple pages of results. For this purpose, we will need to have an *offset* at the client side.  Each fetch operation returns logs after the offset and updates the offset.
   1. APIs -- in addition to list the two API functions, we also need to present a client-side pseudo-code to show the usage.
   1. Implementation -- list the Kubernetes API calls we need and a pseudo-code to show how we are going to call them in function `Fetch`.

File: doc/design/argo_workflow_on_sqlflow.md (hunk 1 above)
Comment (weiguoz) [reply doc/design/argo_workflow_on_sqlflow.md:55]: > The `Token` contains `ID` and fetch status such as `logOffset` and `stepId`:
> 
> https://github.com/sql-machine-learning/sqlflow/blob/36066f128f99acc23e47f6fb93e9d5a6081fad5e/pkg/proto/sqlflow.proto#L52-L58

Avoiding too many logs in the response, do we need an `expect_length` to restrict the log size?

File: doc/design/argo_workflow_on_sqlflow.md (hunk 2 above)
Comment (Yancey0623) [reply doc/design/argo_workflow_on_sqlflow.md:15]: [Kubernetes Logs API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#read-log) supports the `sinceTime` parameter to fetch logs from the specified timestamp,  the client side keeps`FetchToken` object which uses this `timestamp` as `logOffset` to fetch logs in an incremental way, https://github.com/sql-machine-learning/sqlflow/issues/1382 is the more detailed survey.

File: doc/design/argo_workflow_on_sqlflow.md (hunk 3 above)
Comment (Yancey0623) [reply doc/design/argo_workflow_on_sqlflow.md:37]: I will do that after polish this design doc.

File: doc/design/argo_workflow_on_sqlflow.md (hunk 1 above)
Comment (Yancey0623) [reply doc/design/argo_workflow_on_sqlflow.md:55]: It's a good point, [Kubernetes Logs API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#read-log) supports `limit-bytes` parameter to limit each request logs bytes, we can do this enhancement in the `Fetch` implement.

**Pull request merged**